spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy

```diff
@@ -0,0 +1,571 @@
+    /// The target’s Uri
+    uri: URI
+}
+
+list BuildTargetIdentifiers {
+    member: BuildTargetIdentifier
+}
+
+/// Build target contains metadata about an artifact (for example library, test, or binary artifact). Using vocabulary of other build tools:
+///
+/// * sbt: a build target is a combined project + config. Example:
+/// * a regular JVM project with main and test configurations will have 2 build targets, one for main and one for test.
+/// * a single configuration in a single project that contains both Java and Scala sources maps to one BuildTarget.
+/// * a project with crossScalaVersions 2.11 and 2.12 containing main and test configuration in each will have 4 build targets.
+/// * a Scala 2.11 and 2.12 cross-built project for Scala.js and the JVM with main and test configurations will have 8 build targets.
+/// * Pants: a pants target corresponds one-to-one with a BuildTarget
+/// * Bazel: a bazel target corresponds one-to-one with a BuildTarget
+///
+/// The general idea is that the BuildTarget data structure should contain only information that is fast or cheap to compute.
+structure BuildTarget {
+    /// The target’s unique identifier
+    @required
+    id: BuildTargetIdentifier
+    /// A human readable name for this target.
+    /// May be presented in the user interface.
+    /// Should be unique if possible.
+    /// The id.uri is used if None.
+    displayName: String
+    /// The directory where this target belongs to. Multiple build targets are allowed to map
+    /// to the same base directory, and a build target is not required to have a base directory.
+    /// A base directory does not determine the sources of a target, see buildTarget/sources. */
+    baseDirectory: URI
+    /// Free-form string tags to categorize or label this build target.
+    /// For example, can be used by the client to:
+    /// - customize how the target should be translated into the client's project model.
+    /// - group together different but related targets in the user interface.
+    /// - display icons or colors in the user interface.
+    /// Pre-defined tags are listed in `BuildTargetTag` but clients and servers
+    /// are free to define new tags for custom purposes.
+    @required
+    tags: BuildTargetTags
+    /// The set of languages that this target contains.
+    /// The ID string for each language is defined in the LSP.
+    @required
+    languageIds: LanguageIds
+    /// The direct upstream build target dependencies of this build target
+    @required
+    dependencies: BuildTargetIdentifiers
+    /// The capabilities of this build target.
+    @required
+    capabilities: BuildTargetCapabilities
+    /// Kind of data to expect in the `data` field. If this field is not set, the kind of data is not specified.
+    @required
+    dataKind: BuildTargetDataKind
+    /// Language-specific metadata about this target.
+    /// See ScalaBuildTarget as an example.
+    data: Json
+}
+
+@enumKind("open")
```

Comment:
there is no such things as "const" values in smithy. Instead, I've codified the distinction between **open enumerations** (which accept all values of the underlying type but expose a predefined set), and **closed enumerations** (which only accept the values that are defined), as a custom smithy trait that I called `enumKind`, defined alongside the jsonRPC protocol.

Reply:
Ah nice, I was just about to come and comment and this and saw your not. Do we actually want this to be `open`? What happens if we do actually get a tag for example that we don't know about. Shouldn't it be closed forcing one of these?

Reply:
> What happens if we do actually get a tag for example that we don't know about. Shouldn't it be closed forcing one of these? 

Well you tell me : do you want to seal your protocol or leave it open for other usecases than Scala and SBT. I think the latter. If you get a tag you don't know about, the party that doesn't know about it would probably fail one way or another. 

Like, what if you have a Kotlin BSP server ? The `"kotlin"` value should be valid as per the protocol, doesn't mean that Metals should be able to handle "kotlin".

Reply:
Yep, this is very much intentionally open. The protocol definition here gives a few known use cases, but any implementor is free to define new values on client or server side, and either side should be able to handle unknown values